File: src/titanic/titanic_model.py
# titanic_model.py
# Author: Rundong Li
# Description: Logistic Regression model for Titanic survival prediction
# This script loads Titanic train and test data, cleans it, trains a model, and prints outputs.

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

print("Starting Titanic survival prediction script...")

# ===============================
# 1. Load Data
# ===============================
train = pd.read_csv("src/data/train.csv")
test = pd.read_csv("src/data/test.csv")

print("\nData loaded successfully.")
print(f"Training data shape: {train.shape}")
print(f"Test data shape: {test.shape}")

# ===============================
# 2. Basic Data Cleaning
# ===============================
print("\nCleaning data...")

# Fill missing Age and Fare values with median
train["Age"].fillna(train["Age"].median(), inplace=True)
test["Age"].fillna(test["Age"].median(), inplace=True)
test["Fare"].fillna(test["Fare"].median(), inplace=True)

# Fill missing Embarked values with mode
train["Embarked"].fillna(train["Embarked"].mode()[0], inplace=True)
test["Embarked"].fillna(test["Embarked"].mode()[0], inplace=True)

# Convert categorical features
le = LabelEncoder()
for col in ["Sex", "Embarked"]:
    train[col] = le.fit_transform(train[col])
    test[col] = le.fit_transform(test[col])

print("Missing values handled and categorical variables encoded.")

# ===============================
# 3. Feature Selection
# ===============================
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]
target = "Survived"

X = train[features]
y = train[target]

print("\nFeatures used:", features)

# ===============================
# 4. Train/Test Split and Model Training
# ===============================
print("\nTraining logistic regression model...")
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

print("Model training complete.")

# ===============================
# 5. Evaluate on Training Data
# ===============================
train_pred = model.predict(X_train)
train_acc = accuracy_score(y_train, train_pred)
print(f"Training accuracy: {train_acc:.4f}")

# ===============================
# 6. Evaluate on Validation Data
# ===============================
val_pred = model.predict(X_val)
val_acc = accuracy_score(y_val, val_pred)
print(f"Validation accuracy: {val_acc:.4f}")

# ===============================
# 7. Predict on Test Data
# ===============================
print("\nPredicting on test dataset...")
test_pred = model.predict(test[features])

# Save results
output = pd.DataFrame({
    "PassengerId": test["PassengerId"],
    "Survived": test_pred
})
output.to_csv("src/data/predictions.csv", index=False)
print("Predictions saved to src/data/predictions.csv")

print("\nTitanic model run complete.")

File: src/r_titanic/titanic_model.R
# Titanic Survival Prediction - R Version

library(readr)
library(dplyr)
library(caret)

cat("Libraries loaded successfully.\n")

# Load data
train <- read_csv("src/data/train.csv")
test  <- read_csv("src/data/test.csv")
cat("Data loaded successfully.\n")

# Handle missing values
train$Age[is.na(train$Age)] <- median(train$Age, na.rm = TRUE)
test$Age[is.na(test$Age)]   <- median(test$Age, na.rm = TRUE)
test$Fare[is.na(test$Fare)] <- median(test$Fare, na.rm = TRUE)
train$Embarked[is.na(train$Embarked)] <- "S"
test$Embarked[is.na(test$Embarked)]   <- "S"
cat("Missing values handled.\n")

# Convert categorical variables
train$Sex <- as.factor(train$Sex)
test$Sex  <- as.factor(test$Sex)
train$Embarked <- as.factor(train$Embarked)
test$Embarked  <- as.factor(test$Embarked)
cat("Categorical variables converted.\n")

# Logistic regression model
model <- glm(Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare + Embarked,
             data = train, family = binomial)
cat("Model trained successfully.\n")

# Training accuracy
train_pred <- ifelse(predict(model, type = "response") > 0.5, 1, 0)
train_acc <- mean(train_pred == train$Survived)
cat("Training accuracy:", round(train_acc, 4), "\n")

# Predict test set
test_pred <- ifelse(predict(model, newdata = test, type = "response") > 0.5, 1, 0)

# Save results
pred <- data.frame(PassengerId = test$PassengerId, Survived = test_pred)
write_csv(pred, "src/data/predictions_r.csv")
cat("Predictions saved to src/data/predictions_r.csv\n")
